include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce.h

```diff
@@ -288,7 +358,15 @@ struct __parallel_transform_reduce_impl
     }
 }; // struct __parallel_transform_reduce_impl
 
-// General version of parallel_transform_reduce. Calls optimized kernels.
+// General version of parallel_transform_reduce. Processes elements in order to enable non-commutative algorithms.
```

Comment:
I do not quite understand the meaning of "Processes elements in order to enable non-commutative algorithms". Does it mean that this implementation provides a deterministic result when `_ReduceOp` predicate is non-commutative?

It may use `sycl::reduce_over_group` which has the following note:
> The result of a call to these functions is non-deterministic if the binary operator is not commutative and associative

Reply:
Exactly. This version does not require commutative operators. I tried to clarify the comment.

`sycl::reduce_over_group` is used for the group reduction if the type is arithmetic and the operator is commutative (cf. `__has_known_identity`). We process elements in order otherwise.

Reply:
Thanks for the clarification! 

The following is not related to the PR and thus does not block it. I was not able to find a statement that the predicates with available known identities are both associative and commutative. `sycl::minimum`, one of these predicates, returns
the first argument when the arguments are equivalent, so the result for `sycl::minimum(a, b)` and `sycl::minimum(b, a)` are different when `a == b`. Could you open an issue if you think it is a valid case?

Reply:
I agree with @danhoeflinger. The spec for `sycl::reduce_over_group` (https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_reduce) specifies the binary operators in https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:function-objects which include `sycl::minimum` and `sycl::maximum`. The only problem I see is with algorithms that use minimum or maximum but the order matters such as for `min_element()` where the first index with that value needs to be returned. Our implementation uses a user-defined type then and we fall back to an in-order group reduction.

Reply:
Let me make a minor correction to what I said:  for `sycl::minimum` and `sycl::maximum` it `has_known_identity` specifically for integral types, floating point types and sycl::half rather than all arithmetic types.  This leaves out complex types (which is good).